pkg/registry/common/custom/server.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package custom
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
+)
+
+// RegFunction - operation to perform on NSE registration.
+type RegFunction func(ctx context.Context, request *registry.NSERegistration) (*registry.NSERegistration, error)
+
+// RemoveFunction - operation to perform on NSE registration.
+type RemoveFunction func(ctx context.Context, request *registry.RemoveNSERequest) (*registry.RemoveNSERequest, error)
+
+type customRegServer struct {
+	serialize.Executor
+	before RegFunction
```

Comment:
Why not just write a normal chain of chain elements?

Reply:
Because of Bulk NSE stream API, it a lot of code for every chain element and cause errors for every of them, missing some stuff. Probable name is not good, and should be some nseRegFactory probable.

Reply:
Hmm... ok... so basically the API has brokenness... feeds into the work @denis-tingajkin is doing refactoring API.

Reply:
Yes, left impact it for now.